File: core/src/test/java/chaij/BaseExpectationTests.java
package chaij;


import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.sameInstance;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;

@SuppressWarnings("JUnitTestMethodWithNoAssertions")
public class BaseExpectationTests {
	
	@Rule
	public final ExpectedException e = ExpectedException.none();
	
	
	@Test
	public void testLinkers() {
		
		BaseExpectation<?> expectation = new DummyExpectation(null);
		
		assertThat("Linkers should not return a different object.",
		           expectation.to.be.been.is.that.which.and.has.have.with.at.of.same,
		           is(sameInstance(expectation))
		);
	}
	
	
	@Test
	public void testPass() {
		
		BaseExpectation<?> expectation = mock(BaseExpectation.class);
		
		expectation.test(true, "This should not fail.", "But it did!");
	}
	
	
	@Test
	public void testFail() {
		
		e.expect(UnmetExpectationException.class);
		e.expectMessage("This should fail. And it did!");
		BaseExpectation<?> expectation = mock(BaseExpectation.class);
		
		expectation.test(false, "This should fail.", "And it did!");
	}
	
	
	@Test
	public void testNotPass() {
		
		BaseExpectation<?> expectation = mock(BaseExpectation.class);
		
		expectation.not().test(false, "This should not fail.", "But it did!");
	}
	
	
	@Test
	public void testNotNotFail() {
		
		e.expect(UnmetExpectationException.class);
		e.expectMessage("This should fail. And it did!");
		BaseExpectation<?> expectation = mock(BaseExpectation.class);
		
		expectation.not().not().test(false, "This should fail.", "And it did!");
	}
	
	
	@Test
	public void testNotFail() {
		
		e.expect(UnmetExpectationException.class);
		e.expectMessage("This should fail. not And it did!");
		BaseExpectation<?> expectation = mock(BaseExpectation.class);
		
		expectation.not().test(true, "This should fail.", "And it did!");
	}
	
	
	@Test
	public void testWithExplicitMessage() {
		
		e.expect(UnmetExpectationException.class);
		e.expectMessage("Custom message: This should fail. And it did!");
		BaseExpectation<?> expectation = mock(BaseExpectation.class);
		
		expectation.test(false, "Custom message", "This should fail.", "And it did!");
	}
	
	
	@Test
	public void testWithSecondPartSupplier() {
		
		e.expect(UnmetExpectationException.class);
		e.expectMessage("Custom message: This should fail. And it did!");
		BaseExpectation<?> expectation = mock(BaseExpectation.class);
		
		expectation.test(false, "Custom message", "This should fail.", () -> "And it did!");
	}
	
	
	@Test
	public void testWithImplicitMessage() {
		
		e.expect(UnmetExpectationException.class);
		e.expectMessage("Custom message: This should fail. And it did!");
		BaseExpectation<?> expectation = new DummyExpectation("Custom message");
		
		expectation.test(false, "This should fail.", "And it did!");
	}
	
	
	@Test
	public void testBehaviourOnContinueAfterFail() {
		
		BaseExpectation<?> expectation = new DummyExpectation(null);
		e.expect(UnmetExpectationException.class);
		e.expectMessage("Meh. Why?!");
		
		ExceptionReporter.runMultipleAndReport(() -> {
			BaseExpectation<?> other = expectation.test(false, "Meh.", "Why?!");
			
			assertThat("The expectation should be returned.",
			           expectation,
			           is(sameInstance(other))
			);
		});
	}
	
	
	private static final class DummyExpectation extends BaseExpectation<DummyExpectation> {
		
		DummyExpectation(String msg) {super(msg);}
	}
}

File: core/src/test/scala/chaij/BaseExpectationTests.scala
package chaij

import org.scalatest._
import org.mockito.Mockito.{ mock => mockIt }

class BaseExpectationTests extends FlatSpec with Matchers {
  private class DummyExpectation(msg: String) extends BaseExpectation[DummyExpectation](msg)
  private def dummy(message: String = null): DummyExpectation = new DummyExpectation(message)
  private def mock(): DummyExpectation = mockIt(classOf[DummyExpectation])

  "The basic expectation" should "return itself on the normal linkers" in {
    val expectation = dummy()
    expectation
      .to.be.been.is
      .that.which.and.has
      .have.`with`.at.of
      .same should be theSameInstanceAs expectation
  }

  it should "do nothing on a passing test" in {
    mock().test(true, "This should not fail.", "But it did!")
  }

  it should "simply throw on a failing test" in {
    the [UnmetExpectationException] thrownBy {
      mock().test(false, "This should fail.", "And it did!")
    } should have message "This should fail. And it did!"
  }

  it should "respect the not flag on a failing test" in {
    mock().not().test(false, "This should not fail.", "But it did!")
  }

  it should "respect the not flag on a passing test" in {
    the [UnmetExpectationException] thrownBy {
      mock().not().test(true, "This should fail. Here come's a not: '", "'. And that was the not.")
    } should have message "This should fail. Here come's a not: ' not '. And that was the not."
  }

  it should "revert to old behaviour on doubly not" in {
    mock().not().not().test(true, "This should not fail.", "But it did!`")
  }

  it should "use a custom message in failing tests" in {
    the [UnmetExpectationException] thrownBy {
      mock().test(false, "Custom message", "This should fail.", "And it did!")
    } should have message "Custom message: This should fail. And it did!"
  }

  it should "use a message from the expectation itself when failing" in {
    the [UnmetExpectationException] thrownBy {
      dummy("Custom message").test(false, "This should fail.", "And it did!")
    } should have message "Custom message: This should fail. And it did!"
  }

  it should "work with a lazy second in failing tests" in {
    the [UnmetExpectationException] thrownBy {
      mock().test(false, "Custom message", "This should fail.", () => "And it did!")
    } should have message "Custom message: This should fail. And it did!"
  }

  it should "respect the continue after fail behaviour by the ExceptionReporter" in {
    val expectation = dummy()
    the [UnmetExpectationException] thrownBy {
      ExceptionReporter.runMultipleAndReport { () =>
        val afterTest = expectation.test(false, "A", "B")

        afterTest should be theSameInstanceAs expectation
      }
      mock().test(false, "Custom message", "This should fail.", () => "And it did!")
    } should have message "A B"
  }
}
